packages/cli/test/snapshot-matcher.js

```diff
@@ -0,0 +1,205 @@
+
+'use strict';
+
+/*
+A lightweight helper for snapshot-based testing.
+- Feature inspired by https://jestjs.io/docs/en/snapshot-testing
+- Implementation inspired by node-tap
+
+This is an initial experimental version. In the (near) future, we should
+move this file to a standalone package so that all Mocha users can use it.
+*/
+
+const assert = require('assert');
+const path = require('path');
+
+module.exports = {
+  initializeSnapshots,
+};
+
+/**
+ * Create a function to match the given value against a pre-recorder snapshot.
+ *
+ * Example usage:
+ *
+ * ```js
+ * // At the top of your test file, initialize the matcher function.
+ * // You can also move this code to a shared file require()d from tests.
+ * const expectToMatchSnapshot = initializeSnapshots(
+ *   path.resolve(__dirname, '../../__snapshots__'),
+ * );
+ *
+ * // Inside your tests, call `expectToMatchSnapshot(actualValue)`
+ * describe('my feature', () => {
+ *   it('works for strings', function() {
+ *     expectToMatchSnapshot('foo\nbar');
+ *   });
+ * });
+ * ```
+ */
+function initializeSnapshots(snapshotDir) {
+  let currentTest;
+
+  beforeEach(function setupSnapshots() {
+    // eslint-disable-next-line no-invalid-this
+    currentTest = this.currentTest;
+    currentTest.__snapshotCounter = 1;
+  });
+
+  if (!process.env.UPDATE_SNAPSHOTS) {
+    return function expectToMatchSnapshot(actual) {
+      matchSnapshot(snapshotDir, currentTest, actual);
+    };
+  }
+
+  const snapshots = Object.create(null);
+  after(function updateSnapshots() {
+    for (const f in snapshots) {
+      const snapshotFile = buildSnapshotFilePath(snapshotDir, f);
+      writeSnapshotData(snapshotFile, snapshots[f]);
+    }
+  });
+
+  return function expectToRecordSnapshot(actual) {
+    recordSnapshot(snapshots, currentTest, actual);
+  };
+}
+
+function matchSnapshot(snapshotDir, currentTest, actualValue) {
+  assert(
+    typeof actualValue === 'string',
+    'Snapshot matcher supports string values only, but was called with ' +
+      typeof actualValue,
+  );
+
+  const snapshotFile = buildSnapshotFilePath(snapshotDir, currentTest.file);
+  const snapshotData = loadSnapshotData(snapshotFile);
+  const key = buildSnapshotKey(currentTest);
+
+  if (!(key in snapshotData)) {
+    throw new Error(
+      `No snapshot found for ${JSON.stringify(key)}.\n` +
+        'Run the tests with UPDATE_SNAPSHOTS=1 in the environment ' +
+        'to create and update snapshot files.',
+    );
+  }
+
+  assert.deepStrictEqual(actualValue, snapshotData[key]);
+}
+
+function recordSnapshot(snapshots, currentTest, actualValue) {
+  assert(
+    typeof actualValue === 'string',
+    'Snapshot matcher supports string values only, but was called with ' +
+      typeof actualValue,
+  );
+
+  const key = buildSnapshotKey(currentTest);
+  const testFile = currentTest.file;
+  if (!snapshots[testFile]) snapshots[testFile] = Object.create(null);
+  snapshots[testFile][key] = actualValue;
```

Comment:
Nitpick: Would it be better to keep snapshots in separate files to make it more readable?

Reply:
When I first heard about snapshot tests, it was my first question - why are all frameworks keeping multiple snapshots in the same file? I would use a single file for each snapshot too!

Here are few reasons why it may be better to not keep snapshots in separate files:

- Ideally, snapshots should be used for unit-level tests where the snapshots are small, let's say 10-15 lines of code. In which case I feel it would be wasteful to have 20 files with 15 lines of code each, instead of a single file with ~300 LOC.

- I think there is also performance aspect in play. I did not measure it, but I expect that for large test suites like ours, the difference between reading 10 or 200 snapshot files can be noticeable.

- When there is 1:1 relation between a test file (e.g. `controller.integration.js`) and a snapshot file (e.g. `controller.integration.snapshots.js`), it's easier to navigate between these two files. I am not sure how often that's needed though, because snaphots are typically edited automatically by the test framework. 🤷‍♂ 

Because I'd like us to eventually upgrade from Mocha to Jest, I prefer to keep our snapshot files as close to what Jest produces as feasible. That way the migration won't have to touch all snapshot files, it might be super difficult to review such change to catch any unwanted changes!

Reply:
Fair enough 👍 Good learning!